onboard/nav/obstacleAvoidanceStateMachine.hpp

```diff
@@ -0,0 +1,71 @@
+#ifndef OBSTACLEAVOIDANCESTATEMACHINE_HPP
+#define OBSTACLEAVOIDANCESTATEMACHINE_HPP
+
+#include "rover.hpp"
+
+class StateMachine;
+
+// This class is the representation of different 
+// obstacle avoidance algorithms
+enum class ObstacleAvoidanceAlgorithm
+{
+    SimpleAvoidance
+};
+
+// This class is the base class for the logic of the obstacle avoidance state machine
+class ObstacleAvoidanceStateMachine 
+{
+public:
+    /*************************************************************************/
+    /* Public Member Functions */
+    /*************************************************************************/
+    ObstacleAvoidanceStateMachine( StateMachine* stateMachine_ );
+
+    virtual ~ObstacleAvoidanceStateMachine() {}
+
+    void updateObstacleAngle( double bearing );
+
+    NavState run( Rover* phoebe, const rapidjson::Document& roverConfig );
+
+    bool isTennisBallDetected ( Rover* phoebe );
+
+    bool tennisBallReachable( Rover* phoebe, double distance, double bearing );
+
+    virtual Odometry createAvoidancePoint( Rover* phoebe, const double distance ) = 0;
+
+    virtual NavState executeTurnAroundObs( Rover* phoebe, const rapidjson::Document& roverConfig ) = 0;
+
+    virtual NavState executeDriveAroundObs( Rover* phoebe ) = 0;
+protected:
+    /*************************************************************************/
+    /* Protected Member Variables */
+    /*************************************************************************/
+    // Pointer to rover State Machine to access member functions
+    StateMachine* roverStateMachine;
+
+    // Odometry point used when avoiding obstacles.
+    Odometry mObstacleAvoidancePoint;
+
+    // Initial angle to go around obstacle upon detection.
+    double mOriginalObstacleAngle;
+
+    // Initial angle to go around obstacle upon detection.
+    double mOriginalDistance;
+
+    // bool for consecutive obstacle detections
+    bool mJustDetectedObstacle;
+
+    // Last obstacle angle for consecutive angles
+    double mLastObstacleAngle;
+};
+
+// ObstacleAvoidanceFactory creates an ObstacleAvoidanceStateMachine object based on
+// the inputted obstacle avoidance algorithm. This allows for an an ease of transition 
+// between obstacle avoidance algorithms
+ObstacleAvoidanceStateMachine* ObstacleAvoiderFactory( StateMachine* roverStateMachine, 
+                                                       ObstacleAvoidanceAlgorithm algorithm );
+
+#endif //OBSTACLEAVOIDANCESTATEMACHINE_HPP
+
+// TODO: 
+//  remove protected. Bad practice to use protected ( they could simply be public instead )
```

Comment:
this should not be a `TODO`. If this **needs** to be fixed, then now is the time to fix it. I feel that it does not need to be changed and the protected is okay here since only derived classes will use these members so there is no need to make it public and only MRover will be using this so we don't need to make getters/setters